x-pack/plugins/fleet/server/services/agent_policy.ts

```diff
@@ -531,6 +533,166 @@ class AgentPolicyService {
+      agentPolicy: Partial<AgentPolicy>;
+      error: Error | SavedObjectError;
+    }> = [];
+
+    await pMap(agentPolicies, async (agentPolicy) => {
+      try {
+        this.checkTamperProtectionLicense(agentPolicy);
+        await this.checkForValidUninstallToken(agentPolicy, agentPolicy.id);
+        const existingAgentPolicy = existingAgentPolicies.items.find(
+          (i) => i.id === agentPolicy.id
+        );
+        if (!existingAgentPolicy) {
+          throw new AgentPolicyNotFoundError('Agent policy not found'); // This should never happen
+        }
+        if (agentPolicy?.is_protected && !policyHasEndpointSecurity(existingAgentPolicy)) {
+          logger.info(
+            'Agent policy requires Elastic Defend integration to set tamper protection to true'
+          );
+          // force agent policy to be false if elastic defend is not present
+          agentPolicy.is_protected = false;
+        }
+
+        if (existingAgentPolicy.is_managed && !options?.force) {
+          Object.entries(agentPolicy)
+            .filter(([key]) => !KEY_EDITABLE_FOR_MANAGED_POLICIES.includes(key))
+            .forEach(([key, val]) => {
+              if (!isEqual(existingAgentPolicy[key as keyof AgentPolicy], val)) {
+                throw new HostedAgentPolicyRestrictionRelatedError(`Cannot update ${key}`);
+              }
+            });
+        }
+
+        const { monitoring_enabled: monitoringEnabled } = agentPolicy;
+        const packagesToInstall = [];
+        if (!existingAgentPolicy.monitoring_enabled && monitoringEnabled?.length) {
+          packagesToInstall.push(FLEET_ELASTIC_AGENT_PACKAGE);
+        }
+        if (packagesToInstall.length > 0) {
+          await bulkInstallPackages({
+            savedObjectsClient: soClient,
+            esClient,
+            packagesToInstall,
+            spaceId: options?.spaceId || DEFAULT_SPACE_ID,
+            authorizationHeader: options?.authorizationHeader,
+          });
+        }
+        if (
+          existingAgentPolicy.status === agentPolicyStatuses.Inactive &&
+          agentPolicy.status !== agentPolicyStatuses.Active
+        ) {
+          throw new FleetError(
+            `Agent policy ${agentPolicy.id} cannot be updated because it is ${existingAgentPolicy.status}`
+          );
+        }
+
+        if (options.removeProtection) {
+          logger.info(`Setting tamper protection for Agent Policy ${agentPolicy.id} to false`);
+        }
+
+        if (!options.skipValidation) {
+          await validateOutputForPolicy(
+            soClient,
+            agentPolicy,
+            existingAgentPolicy,
+            getAllowedOutputTypeForPolicy(existingAgentPolicy)
+          );
+        }
+
+        policiesToUpdate.push(agentPolicy);
+      } catch (error) {
+        failedPolicies.push({ agentPolicy, error });
+        return;
+      }
+    });
+
+    const { saved_objects: updatedPolicies } = await soClient.bulkUpdate<AgentPolicySOAttributes>(
```

Comment:
do we need to also throttle the updates to the SOs here using `pMap`? meaning: will `soClient.bulkUpdate()` handle passing in `SO_SEARCH_LIMIT` ? or should you maybe break down the updates into batches of 50 (or some other number) of agent policies at a time?